Mathlib/Order/KrullDimension.lean

```diff
@@ -243,14 +243,81 @@ lemma height_eq_iSup_lt_height (x : α) : height x = ⨆ (y : α) (_  : y < x),
     apply iSup₂_le; intro p hp
     apply le_iSup₂_of_le (p.snoc x (hp ▸ hyx)) (by simp) (by simp)
 
-lemma height_le_coe_iff (x : α) (n : ℕ) :
+lemma height_le_coe_iff {x : α} {n : ℕ} :
     height x ≤ n ↔ (∀ y, y < x → height y < n) := by
   conv_lhs => rw [height_eq_iSup_lt_height, iSup₂_le_iff]
   congr! 2 with y _
   cases height y
   · simp
   · norm_cast
 
+/--
+The height of an element is infinite if there exist series of arbitrary length ending in that
+element.
+-/
+lemma height_eq_top_iff {x : α} :
+    height x = ⊤ ↔ ∀ n, ∃ p : LTSeries α, p.last = x ∧ p.length = n where
+  mp h n := by
+    apply exists_series_of_le_height x (n := n)
+    simp [h]
+  mpr h := by
+    rw [height_eq_iSup_last_eq, iSup_subtype', ENat.iSup_coe_eq_top, bddAbove_def]
+    push_neg
+    intro n
+    obtain ⟨p, hlast, hp⟩ := h (n+1)
+    exact ⟨p.length, ⟨⟨⟨p, hlast⟩, by simp [hp]⟩, by simp [hp]⟩⟩
+
+@[simp] lemma height_eq_zero {x : α} : height x = 0 ↔ IsMin x := by
+  simpa [isMin_iff_forall_not_lt] using height_le_coe_iff (x := x) (n := 0)
+
+protected alias ⟨_, IsMin.height_eq_zero⟩ := height_eq_zero
+
+@[simp] lemma height_bot (α : Type*) [Preorder α] [OrderBot α] : height (⊥ : α) = 0 := by simp
+
+lemma coe_lt_height_iff {x : α} {n : ℕ} (hfin : height x < ⊤) :
+    n < height x ↔ (∃ y, y < x ∧ height y = n) where
+  mp h := by
+    obtain ⟨m, hx : height x = m⟩ := Option.ne_none_iff_exists'.mp hfin.ne_top
+    rw [hx] at h; norm_cast at h
+    obtain ⟨p, hp, hlen⟩ := exists_series_of_height_eq_coe x hx
+    use p ⟨n, by omega⟩
+    constructor
+    · rw [← hp]
+      apply LTSeries.strictMono
+      simp [Fin.last]; omega
+    · exact height_eq_index_of_length_eq_height_last (by simp [hlen, hp, hx]) ⟨n, by omega⟩
+  mpr := fun ⟨y, hyx, hy⟩ =>
+    hy ▸ height_strictMono hyx (lt_of_le_of_lt (height_mono hyx.le) hfin)
+
+lemma height_eq_coe_add_one_iff {x : α} {n : ℕ} :
+    height x = n + 1 ↔ height x < ⊤ ∧ (∃ y < x, height y = n) ∧ (∀ y, y < x → height y ≤ n) := by
+  wlog hfin : height x < ⊤
+  · simp_all
+    exact ne_of_beq_false rfl
+  simp only [hfin, true_and]
+  trans n < height x ∧ height x ≤ n + 1
+  · rw [le_antisymm_iff, and_comm]
+    simp [hfin, ENat.lt_add_one_iff, ENat.add_one_le_iff]
+  · congr! 1
+    · exact coe_lt_height_iff hfin
+    · simpa [hfin, ENat.lt_add_one_iff] using height_le_coe_iff (x := x) (n := n+1)
+
+lemma height_eq_coe_iff {x : α} {n : ℕ} :
+    height x = n ↔
+      height x < ⊤ ∧ (n = 0 ∨ ∃ y < x, height y = n - 1) ∧ (∀ y, y < x → height y < n) := by
```

Comment:
```suggestion
      height x < ⊤ ∧ (n = 0 ∨ ∃ y < x, height y = n - 1) ∧ (∀ y < x, height y < n) := by
```
Does it still need a linebreak with that?

Reply:
Yes, still. And the variant with `coheight` (to be added later) even more.

Are linebreaks worth worrying about that much?

Reply:
No. Was just making sure you did actually need it